www/jsonrpc/server.go

```diff
@@ -0,0 +1,110 @@
+package jsonrpc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pacviewer/jrpc-gateway/jrpc"
+	"net/http"
+	"time"
+
+	"github.com/pactus-project/pactus/util/logger"
+	pactus "github.com/pactus-project/pactus/www/grpc/gen/go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type Server struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	config     *Config
+	server     *http.Server
+	grpcClient *grpc.ClientConn
+	handlers   map[string]func(ctx context.Context, params json.RawMessage) (interface{}, error)
+	logger     *logger.SubLogger
+}
+
+func NewServer(conf *Config, enableAuth bool) *Server {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &Server{
+		ctx:    ctx,
+		cancel: cancel,
+		config: conf,
+		logger: logger.NewSubLogger("_jsonrpc", nil),
+	}
+}
+
+func (s *Server) StartServer(grpcServer string) error {
+	if !s.config.Enable {
+		return nil
+	}
+
+	grpcConn, err := grpc.DialContext(
+		s.ctx,
+		grpcServer,
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to dial server: %w", err)
+	}
+
+	s.grpcClient = grpcConn
+
+	blockchain := pactus.NewBlockchainClient(grpcConn)
+	network := pactus.NewNetworkClient(grpcConn)
+	transaction := pactus.NewTransactionClient(grpcConn)
+	wallet := pactus.NewWalletClient(grpcConn)
+
+	blockchainService := pactus.NewBlockchainJsonRpcService(blockchain)
+	networkService := pactus.NewNetworkJsonRpcService(network)
+	transactionService := pactus.NewTransactionJsonRpcService(transaction)
+	walletService := pactus.NewWalletJsonRpcService(wallet)
+
+	s.handlers = make(map[string]func(ctx context.Context, params json.RawMessage) (interface{}, error))
+
+	jgw := jrpc.NewServer()
+	jgw.RegisterServices(&blockchainService, &networkService, &transactionService, &walletService)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", jgw.HttpHandler)
+	server := &http.Server{
+		Addr:              s.config.Listen,
+		ReadHeaderTimeout: 3 * time.Second,
+		Handler:           mux,
+	}
+	go func() {
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			default:
+				if err := server.ListenAndServe(); err != nil {
+					s.logger.Error("error while establishing JSON-RPC connection", "error", err)
+				}
+			}
+		}
+	}()
+	s.logger.Info("jsonrpc started listening", "address", s.config.Listen)
```

Comment:
It is better to create a listener and then serve the HTTP (line 83). 
Then you can get the exact listening address from the listener, especially if the user sets `0` for the listening port; the listener returns the exact port.

Reply:
How can i do that?
It seems http package only supports ListenAndServe methods.

Reply:
As an example, look [here](https://github.com/pactus-project/pactus/blob/20d5d833394ea4f3b62cf14631b9584b4a9a1a28/www/http/server.go#L87)